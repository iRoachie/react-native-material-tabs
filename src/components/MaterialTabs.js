// @flow

import React from 'react';
import PropTypes from 'prop-types';
import { Animated, ScrollView, View, Text } from 'react-native';
import { Bar, TabTrack } from '../lib/styles';
import values from '../lib/values';
import Tab from './Tab';
import Indicator from './Indicator';
import type { ContentType } from './Tab/Tab';

type Props = {
  allowFontScaling: boolean,
  selectedIndex: number,
  barColor: string,
  barHeight: number,
  tabWidth: number | string,
  activeTextColor: string,
  indicatorColor: string,
  inactiveTextColor: string,
  scrollable: boolean,
  textStyle: any,
  activeTextStyle: any,
  items: ContentType[],
  uppercase: boolean,
  keyboardShouldPersistTaps: string,
  onChange: (index: number) => void,
};

type State = {
  tabWidth: number,
  barWidth: number,
  indicatorPosition: Animated.Value,
};

const getKeyForTab = (item: ContentType) =>
  typeof item === 'string' ? item : item.key;

export default class MaterialTabs extends React.Component<Props, State> {
  static propTypes = {
    allowFontScaling: PropTypes.bool,
    selectedIndex: PropTypes.number,
    barColor: PropTypes.string,
    barHeight: PropTypes.number,
    tabWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    activeTextColor: PropTypes.string,
    indicatorColor: PropTypes.string,
    inactiveTextColor: PropTypes.string,
    scrollable: PropTypes.bool,
    textStyle: Text.propTypes.style,
    activeTextStyle: Text.propTypes.style,
    items: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.element])
    ).isRequired,
    uppercase: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    keyboardShouldPersistTaps: PropTypes.string,
  };

  static defaultProps = {
    allowFontScaling: true,
    selectedIndex: 0,
    barColor: '#13897b',
    barHeight: values.barHeight,
    tabWidth: values.tabWidth,
    activeTextColor: '#fff',
    indicatorColor: '#fff',
    inactiveTextColor: 'rgba(255, 255, 255, 0.7)',
    scrollable: false,
    textStyle: null,
    uppercase: true,
    activeTextStyle: {},
    keyboardShouldPersistTaps: 'never',
  };

  state = {
    tabWidth: 0,
    barWidth: 0,
    indicatorPosition: new Animated.Value(0),
  };

  shouldComponentUpdate(nextProps: Props) {
    // Prevent scrolling out of bounds
    return (
      nextProps.selectedIndex < nextProps.items.length &&
      nextProps.selectedIndex >= 0
    );
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.items.length !== prevProps.items.length) {
      this.bar &&
        this.bar.measure((_, b, width) => {
          this.getTabWidth(width);
        });
    }

    this.selectTab();
  }

  scrollView: ScrollView | null;

  bar: View | null;

  getAnimateValues() {
    const idx = this.props.selectedIndex;
    const scrollValue = this.state.tabWidth;

    // All props for fixed tabs are the same
    if (!this.props.scrollable) {
      return {
        indicatorPosition: idx === 0 ? 0 : idx * scrollValue,
        scrollPosition: 0,
      };
    }

    switch (idx) {
      case 0: // First tab
        return {
          indicatorPosition: 0,
          scrollPosition: 0,
        };
      case 1: // Second tab
        return {
          indicatorPosition: this.state.barWidth * 0.5 - scrollValue / 4,
          scrollPosition: scrollValue * 0.25,
        };
      case this.props.items.length - 1: // Last tab
        return {
          indicatorPosition:
            scrollValue * (idx - 1) +
            (this.state.barWidth * 0.5 - scrollValue / 4),
          scrollPosition: scrollValue * (idx - 2) + scrollValue * 0.5,
        };
      default:
        // Any tabs between second and last
        return {
          indicatorPosition:
            scrollValue * (idx - 1) +
            (this.state.barWidth * 0.5 - scrollValue / 4),
          scrollPosition: scrollValue * 0.25 + scrollValue * (idx - 1),
        };
    }
  }

  getTabWidth(width: number) {
    const userDefinedWidth =
      this.props.tabWidth &&
      MaterialTabs.parseUserDefinedTabWidth(width, this.props.tabWidth);
    const defaultScrollableTabWidth = width * 0.4;

    const tabWidth = this.props.scrollable
      ? userDefinedWidth || defaultScrollableTabWidth
      : width / this.props.items.length;

    this.setState({
      barWidth: width,
      tabWidth,
    });
  }

  static parseUserDefinedTabWidth(barWidth: number, tabWidth: number | string) {
    if (typeof tabWidth === 'number') {
      return tabWidth;
    }

    if (typeof tabWidth === 'string') {
      const isPercentage =
        tabWidth.length > 1 && tabWidth[tabWidth.length - 1] === '%';

      if (isPercentage) {
        const valueWithoutPercentSymbol = tabWidth.slice(
          0,
          tabWidth.length - 1
        );
        const numericPercentValue = parseFloat(valueWithoutPercentSymbol);
        const numericDecimalValue = numericPercentValue / 100;

        return barWidth * numericDecimalValue;
      }

      return parseFloat(tabWidth);
    }

    throw new Error('Unreachable');
  }

  selectTab() {
    Animated.spring(this.state.indicatorPosition, {
      toValue: this.getAnimateValues().indicatorPosition,
      tension: 300,
      friction: 20,
      useNativeDriver: true,
    }).start();

    if (this.scrollView) {
      this.scrollView.scrollTo({
        x: this.getAnimateValues().scrollPosition,
      });
    }
  }

  renderContent() {
    return (
      <Bar
        ref={(ref: View | null) => {
          this.bar = ref;
        }}
        barColor={this.props.barColor}
        barHeight={this.props.barHeight}
        onLayout={event => this.getTabWidth(event.nativeEvent.layout.width)}
      >
        <ScrollView
          horizontal
          ref={ref => {
            this.scrollView = ref;
          }}
          showsHorizontalScrollIndicator={false}
          keyboardShouldPersistTaps={this.props.keyboardShouldPersistTaps}
          scrollEnabled={this.props.scrollable}
        >
          <TabTrack barHeight={this.props.barHeight}>
            {this.props.items.map((item, idx) => (
              <Tab
                allowFontScaling={this.props.allowFontScaling}
                content={item}
                key={getKeyForTab(item)}
                stretch={!this.props.scrollable}
                onPress={() => this.props.onChange(idx)}
                active={idx === this.props.selectedIndex}
                activeTextColor={this.props.activeTextColor}
                textStyle={this.props.textStyle}
                activeTextStyle={
                  this.props.selectedIndex === idx
                    ? this.props.activeTextStyle
                    : {}
                }
                tabHeight={this.props.barHeight}
                tabWidth={this.state.tabWidth}
                uppercase={this.props.uppercase}
                inActiveTextColor={this.props.inactiveTextColor}
              />
            ))}
          </TabTrack>

          <Indicator
            color={this.props.indicatorColor}
            value={this.state.indicatorPosition}
            tabWidth={this.state.tabWidth}
          />
        </ScrollView>
      </Bar>
    );
  }

  render() {
    return this.props.items ? this.renderContent() : null;
  }
}
