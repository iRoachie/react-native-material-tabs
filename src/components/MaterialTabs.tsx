import React, { useState, useEffect, useCallback } from 'react';
import {
  Animated,
  ScrollView,
  View,
  ScrollViewProps,
  StyleProp,
  TextStyle,
  I18nManager,
} from 'react-native';

import Tab from './Tab';
import Indicator from './Indicator';
import { ContentType } from './Tab/Tab';

import { Bar, TabTrack } from '../lib/styles';

interface Props extends Pick<ScrollViewProps, 'keyboardShouldPersistTaps'> {
  allowFontScaling: boolean;
  selectedIndex: number;
  barColor: string;
  barHeight: number;
  activeTextColor: string;
  indicatorColor: string;
  inactiveTextColor: string;
  scrollable: boolean;
  textStyle: StyleProp<TextStyle>;
  activeTextStyle: StyleProp<TextStyle>;
  items: ContentType[];
  uppercase: boolean;
  onChange(index: number): void;
  indicatorHeight: number;
}

const getKeyForTab = (item: ContentType) =>
  typeof item === 'string' ? item : item.key;

const MaterialTabs: React.FC<Props> = ({
  items,
  selectedIndex,
  scrollable,
  keyboardShouldPersistTaps,
  barHeight,
  onChange,
  allowFontScaling,
  activeTextColor,
  textStyle,
  activeTextStyle,
  inactiveTextColor,
  uppercase,
  indicatorColor,
  barColor,
  indicatorHeight,
}) => {
  const [tabWidth, setTabWidth] = useState(0);
  const [barWidth, setBarWidth] = useState(0);
  const [indicatorPosition] = useState(new Animated.Value(0));
  const scrollView = React.createRef<ScrollView>();
  const bar = React.createRef<View>();

  const getTabWidth = useCallback(
    (width: number = 0) => {
      if (!scrollable) {
        setTabWidth(width / items.length);
      }

      setBarWidth(width);
    },
    [items.length, scrollable]
  );

  useEffect(() => {
    const getAnimateValues = () => {
      const scrollValue = !scrollable ? tabWidth : barWidth * 0.4;

      const indicator = I18nManager.isRTL
        ? -selectedIndex * scrollValue
        : selectedIndex * scrollValue;

      // All props for fixed tabs are the same
      if (!scrollable) {
        return {
          indicatorPosition: indicator,
          scrollPosition: 0,
        };
      }

      return {
        indicatorPosition: indicator,
        scrollPosition: I18nManager.isRTL
          ? scrollValue * 0.25 +
            scrollValue * (items.length - selectedIndex - 2)
          : scrollValue * 0.25 + scrollValue * (selectedIndex - 1),
      };
    };

    const selectTab = () => {
      const values = getAnimateValues();

      Animated.spring(indicatorPosition, {
        toValue: values.indicatorPosition,
        tension: 300,
        friction: 20,
        useNativeDriver: true,
      }).start();

      if (scrollView.current) {
        scrollView.current.scrollTo({
          x: values.scrollPosition,
        });
      }
    };

    bar.current &&
      bar.current.measure((_, b, width) => {
        getTabWidth(width);
      });

    selectTab();
  }, [
    bar,
    barWidth,
    getTabWidth,
    indicatorPosition,
    items.length,
    scrollView,
    scrollable,
    selectedIndex,
    tabWidth,
  ]);

  return (
    items && (
      <Bar
        ref={bar}
        barColor={barColor}
        barHeight={barHeight}
        onLayout={event => getTabWidth(event.nativeEvent.layout.width)}
      >
        <ScrollView
          horizontal
          ref={scrollView}
          showsHorizontalScrollIndicator={false}
          keyboardShouldPersistTaps={keyboardShouldPersistTaps}
          scrollEnabled={scrollable}
        >
          <TabTrack barHeight={barHeight} indicatorHeight={indicatorHeight}>
            {items.map((item, idx) => (
              <Tab
                allowFontScaling={allowFontScaling}
                content={item}
                key={getKeyForTab(item) || undefined}
                onPress={() => onChange(idx)}
                active={idx === selectedIndex}
                activeTextColor={activeTextColor}
                textStyle={textStyle}
                activeTextStyle={selectedIndex === idx && activeTextStyle}
                tabHeight={barHeight}
                tabWidth={!scrollable ? tabWidth : barWidth * 0.4}
                uppercase={uppercase}
                inActiveTextColor={inactiveTextColor}
              />
            ))}
          </TabTrack>

          <Indicator
            color={indicatorColor}
            value={indicatorPosition}
            tabWidth={!scrollable ? tabWidth : barWidth * 0.4}
            height={indicatorHeight}
          />
        </ScrollView>
      </Bar>
    )
  );
};

MaterialTabs.defaultProps = {
  allowFontScaling: true,
  selectedIndex: 0,
  barColor: '#13897b',
  barHeight: 48,
  activeTextColor: '#fff',
  indicatorColor: '#fff',
  inactiveTextColor: 'rgba(255, 255, 255, 0.7)',
  scrollable: false,
  uppercase: true,
  keyboardShouldPersistTaps: 'never',
  indicatorHeight: 2,
};

export default MaterialTabs;
