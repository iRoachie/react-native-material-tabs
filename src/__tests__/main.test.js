import React from 'react';
import { Text } from 'react-native';
import Adapter from 'enzyme-adapter-react-16';
import { create } from 'react-test-renderer';
import { shallow, configure } from 'enzyme';

import MaterialTabs from '../index';

configure({ adapter: new Adapter() });

let component;
let onChange;

describe('Main', () => {
  beforeAll(() => {
    // eslint-disable-next-line no-console
    console.warn = jest.fn();
  });

  beforeEach(() => {
    onChange = jest.fn();
    component = (
      <MaterialTabs
        selectedIndex={0}
        items={['Tab1', 'Tab2']}
        onChange={onChange}
      />
    );
  });

  it('should render without errors', () => {
    const tree = create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render scrollable without errors', () => {
    const tree = create(
      <MaterialTabs selectedIndex={0} items={['Tab1', 'Tab2']} scrollable />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render with correct tabs', () => {
    const wrapper = shallow(component);
    const tabs = wrapper.find('Tab');

    expect(tabs.length).toBe(2);
  });

  it('should fire onChange event', () => {
    const wrapper = shallow(component);

    wrapper
      .find('Tab')
      .at(1)
      .props()
      .onPress();

    expect(onChange).toHaveBeenCalledWith(1);
  });

  it('should change tabs', () => {
    const wrapper = shallow(component);
    let tabs = wrapper.find('Tab');

    // Assert initial state
    expect(tabs.at(0).props().active).toBe(true);

    wrapper.setProps({ selectedIndex: 1 });
    tabs = wrapper.find('Tab');

    expect(tabs.at(0).props().active).toBe(false);
    expect(tabs.at(1).props().active).toBe(true);
  });

  it('should apply custom fontFamily to tab', () => {
    const textComponent = (
      <MaterialTabs
        selectedIndex={0}
        items={['Tab1', 'Tab2']}
        onChange={onChange}
        textStyle={{ fontFamily: 'Papyrus' }}
      />
    );

    const wrapper = shallow(textComponent);
    const tree = create(textComponent).toJSON();

    const tab = wrapper
      .find('Tab')
      .at(0) // Tab
      .dive() // TabButton
      .children() // TabBody
      .children(); // TabText

    expect(tab.props().style).toEqual({ fontFamily: 'Papyrus' });
    expect(tree).toMatchSnapshot();
  });

  it('should apply custom activeTextStyle to active tab', () => {
    const textComponent = (
      <MaterialTabs
        selectedIndex={0}
        items={['Tab1', 'Tab2']}
        onChange={onChange}
        activeTextStyle={{ color: 'pink' }}
      />
    );

    const wrapper = shallow(textComponent);
    const tree = create(textComponent).toJSON();

    const tab = wrapper
      .find('Tab')
      .at(0) // Tab
      .dive() // TabButton
      .children() // TabBody
      .children(); // TabText

    expect(tab.props().style).toEqual({ color: 'pink' });
    expect(tree).toMatchSnapshot();
  });

  it('should display tab labels not uppercased', () => {
    const tabs = (
      <MaterialTabs
        selectedIndex={0}
        items={['Tab1', 'Tab2']}
        onChange={onChange}
        uppercase={false}
      />
    );

    const tree = create(tabs).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('when items is an array of react elements', () => {
    it('should render tabs', () => {
      const tabs = (
        <MaterialTabs
          selectedIndex={0}
          items={[<Text key="1">Tab1</Text>, <Text key="2">Tab2</Text>]}
          onChange={onChange}
        />
      );
      const tree = create(tabs).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
