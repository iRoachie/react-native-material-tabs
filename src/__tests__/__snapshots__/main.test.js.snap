// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should apply custom activeTextStyle to active tab 1`] = `
<View
  barColor="#13897b"
  barHeight={48}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#13897b",
        "height": 48,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    keyboardShouldPersistTaps="never"
    scrollEnabled={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        barHeight={48}
        indicatorHeight={2}
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 46,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="#fff"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {
                    "color": "pink",
                  },
                ]
              }
            >
              TAB1
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="rgba(255, 255, 255, 0.7)"
              style={
                Array [
                  Object {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              TAB2
            </Text>
          </View>
        </View>
      </View>
      <View
        color="#fff"
        height={2}
        style={
          Object {
            "backgroundColor": "#fff",
            "bottom": 0,
            "height": 2,
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 0,
          }
        }
        tabWidth={0}
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Main should apply custom fontFamily to tab 1`] = `
<View
  barColor="#13897b"
  barHeight={48}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#13897b",
        "height": 48,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    keyboardShouldPersistTaps="never"
    scrollEnabled={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        barHeight={48}
        indicatorHeight={2}
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 46,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="#fff"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {
                    "fontFamily": "Papyrus",
                  },
                ]
              }
            >
              TAB1
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="rgba(255, 255, 255, 0.7)"
              style={
                Array [
                  Object {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {
                    "fontFamily": "Papyrus",
                  },
                ]
              }
            >
              TAB2
            </Text>
          </View>
        </View>
      </View>
      <View
        color="#fff"
        height={2}
        style={
          Object {
            "backgroundColor": "#fff",
            "bottom": 0,
            "height": 2,
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 0,
          }
        }
        tabWidth={0}
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Main should display tab labels not uppercased 1`] = `
<View
  barColor="#13897b"
  barHeight={48}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#13897b",
        "height": 48,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    keyboardShouldPersistTaps="never"
    scrollEnabled={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        barHeight={48}
        indicatorHeight={2}
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 46,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="#fff"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              Tab1
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="rgba(255, 255, 255, 0.7)"
              style={
                Array [
                  Object {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              Tab2
            </Text>
          </View>
        </View>
      </View>
      <View
        color="#fff"
        height={2}
        style={
          Object {
            "backgroundColor": "#fff",
            "bottom": 0,
            "height": 2,
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 0,
          }
        }
        tabWidth={0}
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Main should render scrollable without errors 1`] = `
<View
  barColor="#13897b"
  barHeight={48}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#13897b",
        "height": 48,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    keyboardShouldPersistTaps="never"
    scrollEnabled={true}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        barHeight={48}
        indicatorHeight={2}
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 46,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="#fff"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              TAB1
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="rgba(255, 255, 255, 0.7)"
              style={
                Array [
                  Object {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              TAB2
            </Text>
          </View>
        </View>
      </View>
      <View
        color="#fff"
        height={2}
        style={
          Object {
            "backgroundColor": "#fff",
            "bottom": 0,
            "height": 2,
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 0,
          }
        }
        tabWidth={0}
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Main should render without errors 1`] = `
<View
  barColor="#13897b"
  barHeight={48}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#13897b",
        "height": 48,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    keyboardShouldPersistTaps="never"
    scrollEnabled={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        barHeight={48}
        indicatorHeight={2}
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 46,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="#fff"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              TAB1
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              allowFontScaling={true}
              color="rgba(255, 255, 255, 0.7)"
              style={
                Array [
                  Object {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "minWidth": "100%",
                    "textAlign": "center",
                  },
                  Object {},
                ]
              }
            >
              TAB2
            </Text>
          </View>
        </View>
      </View>
      <View
        color="#fff"
        height={2}
        style={
          Object {
            "backgroundColor": "#fff",
            "bottom": 0,
            "height": 2,
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 0,
          }
        }
        tabWidth={0}
      />
    </View>
  </RCTScrollView>
</View>
`;

exports[`Main when items is an array of react elements should render tabs 1`] = `
<View
  barColor="#13897b"
  barHeight={48}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#13897b",
        "height": 48,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    keyboardShouldPersistTaps="never"
    scrollEnabled={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        barHeight={48}
        indicatorHeight={2}
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 46,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              style={
                Array [
                  undefined,
                  Object {
                    "color": "#fff",
                  },
                ]
              }
            >
              Tab1
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "width": 0,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 48,
                  "justifyContent": "center",
                  "paddingLeft": 12,
                  "paddingRight": 12,
                },
              ]
            }
            tabHeight={48}
          >
            <Text
              style={
                Array [
                  undefined,
                  Object {
                    "color": "rgba(255, 255, 255, 0.7)",
                  },
                ]
              }
            >
              Tab2
            </Text>
          </View>
        </View>
      </View>
      <View
        color="#fff"
        height={2}
        style={
          Object {
            "backgroundColor": "#fff",
            "bottom": 0,
            "height": 2,
            "position": "absolute",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": 0,
          }
        }
        tabWidth={0}
      />
    </View>
  </RCTScrollView>
</View>
`;
